fvm/state/state.go

```diff
@@ -233,3 +234,56 @@ func addressFromOwner(owner string) (flow.Address, bool) {
 	address := flow.BytesToAddress(ownerBytes)
 	return address, true
 }
+
+// IsFVMStateKey returns true if the
+// key is controlled by the fvm env and
+// return false otherwise (key controlled by the cadence env)
+func IsFVMStateKey(owner, controller, key string) bool {
+
+	// check if is a service level key (owner and controller is empty)
+	// cases:
+	// 		- "", "", "uuid"
+	// 		- "", "", "account_address_state"
+	if len(owner) == 0 && len(controller) == 0 {
+		return true
+	}
+	// check account level keys
+	// cases:
+	// 		- address, address, "public_key_%d" (index)
+	// 		- address, address, "public_key_count"
+	// 		- address, address, "contract_names"
+	// 		- address, address, "code.%s" (contract name)
+	// 		- address, "", exists
+	// 		- address, "", "storage_used"
+	// 		- address, "", "frozen"
+
+	if owner == controller && key == KeyPublicKeyCount {
+		return true
+	}
+
+	if owner == controller && bytes.HasPrefix([]byte(key), []byte("public_key_")) {
+		return true
+	}
+
+	if owner == controller && key == KeyContractNames {
+		return true
+	}
+
+	if owner == controller && bytes.HasPrefix([]byte(key), []byte(KeyCode)) {
+		return true
+	}
+
+	if len(controller) == 0 && key == KeyExists {
+		return true
+	}
+
+	if len(controller) == 0 && key == KeyStorageUsed {
+		return true
+	}
+
+	if len(controller) == 0 && key == KeyAccountFrozen {
+		return true
+	}
```

Comment:
Maybe factor out the common parts, i.e. the `owner == controller` cases, and the `len(controller) == 0` cases, so they don't have to be re-tested

Reply:
Applied at 33a109f